### PR TITLE
Release 1.2.2

### PR DESCRIPTION
**Pull request by KGuillemot**
### Improvements
 - Portal : Add API + automatically create OpenID service if IDP portal
 - Darwin : Add **vast**/**vaml** darwin filters, with according Rsyslog configurations
 - Portal templates : Add possibility to add custom images + improve edition dashboard
 - Portal : Set external_listener as Frontend instead of listener, to permit IDP listening domain on existing frontend
 - Api Cybereason : Add more informational loggers
 - Reputation contexts : Add enable_hour_download config, to download database once
 - OpenID : Add 'verify_certificate' and 'use_proxy' attributes for configuration downloading
 - Ldap : Remove userPassword retrieve from ldap
 - API Symantec : Improve parsing loop for very big amount of logs
 
### Fixes
 - Ldap client : Fix API
 - Install : Fix migration error when bootstrapping Vulture
 - HAProxy : Fix portal, workflow and frontend configuration templates
 - Ldap client : Fix error in group_search method
 - Error templates : Add http headers at beginning of templates, for Haproxy compatibility
 - API SentinelOne : Use system proxy in api requests
 - IDP : Fix IDP update/delete
 - Frontend : Fix API for non-json post data
 - Frontend : Fix custom http headers configuration
 - Portal : Fix OpenID token creation in /oauth2/authorize endpoint
 - Portal : Fix OpenID responders + SSO forward
 - Portal : Fix custom scope re-edition
 - Portal : Fix Apache configuration for Authorization header
 - Portal : Fix 503 error for OAuth2 callback responder
 - Haproxy : Remove X-Forwarded-* headers from frontend, set them in portal backends, to prevent conflict with custom headers

### CODE REVIEW

File: vulture_os/portal/system/authentications.py [hunk 1]
```diff
@@ -155,73 +161,68 @@ def authenticate_on_backend(self, backend):
...
+        for backend in self.workflow.authentication.repositories.exclude(subtype="openid"):
+            try:
+                authentication_results = self.authenticate_on_backend(backend)
+                self.backend_id = str(backend.id)
+                logger.info("AUTH::authenticate: User '{}' successfully authenticated on fallback backend '{}'"
+                            .format(self.credentials[0], backend))
+                return authentication_results
+
+            except (AuthenticationError, ACLError, DBAPIError, PyMongoError, LDAPError) as e:
+                logger.error("AUTH::authenticate: Authentication failure for username '{}' on fallback backend '{}'"
```
Comment (frikilax) [inline vulture_os/portal/system/authentications.py:174]:
```suggestion
                logger.info("AUTH::authenticate: User '{}' successfully authenticated on backend '{}'"
                            .format(self.credentials[0], backend))
                return authentication_results

            except (AuthenticationError, ACLError, DBAPIError, PyMongoError, LDAPError) as e:
                logger.error("AUTH::authenticate: Authentication failure for username '{}' on backend '{}'"
```

File: vulture_os/portal/system/authentications.py [hunk 2]
```diff
@@ -155,73 +161,68 @@ def authenticate_on_backend(self, backend):
...
                                                                     self.workflow.authentication.otp_repository)
             logger.debug("AUTH::register_sso: DoubleAuthentication required : "
                          "successfully written in Redis for user '{}'".format(username))
-        return portal_cookie, oauth2_token
+        return portal_cookie, self.oauth2_token
+
+    def register_openid(self, openid_token, **kwargs):
+        # Generate a new OAuth2 token
+        #self.oauth2_token = Uuid4().generate()
+        # Register it into session
```
Comment (frikilax) [inline vulture_os/portal/system/authentications.py:222]:
```suggestion
        # Register it into session using existing oauth2 token
```

File: vulture_os/portal/views/logon.py [hunk 3]
```diff
@@ -170,6 +208,7 @@ def openid_callback(request, workflow_id, repo_id):
 
         # Create user scope depending on GUI configuration attributes
         user_scope = workflow.authentication.get_user_scope(claims, repo_attributes)
+        logger.info(user_scope)
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:211]:
```suggestion
        logger.debug(user_scope)
```

File: vulture_os/portal/views/logon.py [hunk 4]
```diff
@@ -408,166 +265,242 @@ def openid_authorize(request, portal_id):
...
-        # Redirect to the same uri, to stock token in Redis via session filter
-        return HttpResponseRedirect("")
+    try:
+        assert request.headers.get('Authorization')
+        assert request.headers.get('Authorization').startswith("Bearer ")
+
+        oauth2_token = request.headers.get('Authorization').replace("Bearer ", "")
+        session = REDISOauth2Session(REDISBase(), f"oauth2_{oauth2_token}")
+        assert session['scope']
+        logger.info(session['scope'])
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:378]:
```suggestion
        logger.debug(session['scope'])
```

File: vulture_os/portal/views/logon.py [hunk 5]
```diff
@@ -408,166 +265,242 @@ def openid_authorize(request, portal_id):
...
+        # If the user is not authenticated and application need authentication
+        if not authentication.is_authenticated():
+            try:
+                backend_id = authentication.authenticate_sso_acls()
+                if not backend_id:
+                    # Retrieve credentials
+                    authentication.retrieve_credentials(request)
+                    logger.debug("PORTAL::log_in: Credentials successfully retrieved")
+
+                    # Authenticate user with credentials retrieven
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:439]:
```suggestion
                    # Authenticate user with retrieved credentials
```

File: vulture_os/portal/views/logon.py [hunk 6]
```diff
@@ -408,166 +265,242 @@ def openid_authorize(request, portal_id):
...
+                    # Register authentication results in Redis
+                    portal_cookie, oauth2_token = authentication.register_user(authentication_results)
+                    logger.debug("PORTAL::log_in: User {} successfully registered in Redis".format(authentication.credentials[0]))
+
+                    if authentication_results.get('password_expired', None):
+                        logger.info("PORTAL::log_in: User '{}' must change its password, redirect to self-service portal"
+                                    .format(authentication.credentials[0]))
+                        app_url = workflow.get_redirect_uri()
+                        return HttpResponseRedirect(str(token_name)+'/self/change')
+                # If the user is already authenticated (retrieven with RedisPortalSession ) => SSO
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:452]:
```suggestion
                # If the user is already authenticated (retrieved with RedisPortalSession ) => SSO
```

File: vulture_os/portal/views/logon.py [hunk 7]
```diff
@@ -408,166 +265,242 @@ def openid_authorize(request, portal_id):
...
+                                 .format(authentication.credentials[0], e))
+                return authentication.ask_credentials_response(request=request)
+
+            except REDISWriteError as e:
+                logger.error("PORTAL::log_in: RedisWriteError while trying to register user '{}' informations : {}"
+                             .format(authentication.credentials[0], e))
+                return HttpResponseServerError()
+
+            except Exception as e:
+                logger.error("ERROR !")
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:492]:
```suggestion
```

Comment (frikilax) [reply vulture_os/portal/views/logon.py:492]: useless as it is

File: vulture_os/portal/views/logon.py [hunk 8]
```diff
@@ -656,85 +586,125 @@ def log_in(request, workflow_id=None):
...
+                # Use 'sso_data' and 'url' to authenticate user on application
+                response = sso_forward.authenticate(sso_data, post_url=url, redis_session=authentication.redis_portal_session)
+                logger.info("PORTAL::log_in: SSOForward performing success")
+                # Generate response depending on application.sso_forward options
+                final_response = sso_forward.generate_response(request, response, authentication.get_redirect_url())
+                logger.info("PORTAL::log_in: SSOForward response successfuly generated")
+
+                return final_response
+
+            # If learning credentials cannot be retrieven : ask them
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:640]:
```suggestion
            # If learning credentials cannot be retrieved : ask them
```

File: vulture_os/portal/views/logon.py [hunk 9]
```diff
@@ -656,85 +586,125 @@ def log_in(request, workflow_id=None):
...
+            except CredentialsMissingError as e:
+                logger.error("PORTAL::log_in: Learning credentials missing : asking-them")
+                return authentication.ask_learning_credentials(request=request,
+                                                               fields=e.fields_missing)
+
+            # If KerberosBackend object cannot be retrieven from mongo with the backend_id that the user is authenticated on
```
Comment (frikilax) [inline vulture_os/portal/views/logon.py:646]:
```suggestion
            # If KerberosBackend object cannot be retrieved from mongo with the backend_id that the user is authenticated on
```

File: vulture_os/services/frontend/models.py [hunk 10]
```diff
@@ -1477,6 +1478,13 @@ def rsyslog_only_conf(self):
         """ Check if this frontend has only rsyslog configuration, not haproxy at all """
         return self.mode == "impcap" or (self.mode == "log" and (self.listening_mode in ("udp", "file", "api", "kafka", "redis")))
 
+    @property
+    def has_tls(self):
+        for listener in self.listener_set.all():
+            if listener.is_tls:
+                return True
+        return False
+
```
Comment (frikilax) [inline vulture_os/services/frontend/models.py:1487]:
```suggestion
```
seems unused

File: vulture_os/services/frontend/models.py [hunk 11]
```diff
@@ -1589,6 +1601,9 @@ def save(self, *args, **kwargs):
                 pass
         super().save(*args, **kwargs)
 
+    def get_nodes(self):
+        return [nic.node for nic in self.network_address.nic.all()]
```
Comment (frikilax) [inline vulture_os/services/frontend/models.py:1605]: shouldn't that be better with a set() ?

File: vulture_os/toolkit/auth/ldap_client.py [hunk 12]
```diff
@@ -254,7 +275,7 @@ def _search(self, dn, ldap_query, username, attr_list=None):
             if not controls[0].cookie:
                 break
             page_control.cookie = controls[0].cookie
-        logger.info("LDAP search_s result is: {}".format(result))
+        #logger.info("LDAP search_s result is: {}".format(result))
```
Comment (frikilax) [inline vulture_os/toolkit/auth/ldap_client.py:278]:
```suggestion

```

File: vulture_os/authentication/openid/models.py [hunk 13]
```diff
@@ -195,13 +205,16 @@ def get_client(self):
...
         # TODO : Handle CA_BUNDLE
         # If loaded data is too old, reload it again
         refresh_time = timezone.now() - timedelta(hours=CONFIG_RELOAD_INTERVAL)
         if (self.last_config_time is None or self.last_config_time < refresh_time)\
                 or test:
-            r = requests.get("{}/.well-known/openid-configuration".format(self.provider_url), proxies=get_proxy())
+            logger.info(get_proxy() if self.use_proxy else None)
+            r = requests.get("{}/.well-known/openid-configuration".format(self.provider_url),
+                             proxies=get_proxy() if self.use_proxy else None,
+                             verify=self.verify_certificate)
```
[the remark this answers is not in the document]
Comment (KGuillemot) [reply vulture_os/authentication/openid/models.py]: Thanks, added.

File: vulture_os/portal/system/authentications.py (hunk 2 above)
Comment (KGuillemot) [reply vulture_os/portal/system/authentications.py:222]: I prefer to keep this line, as a reminder for later improvement - we should regenerate a new token each time

File: vulture_os/portal/views/logon.py (hunk 3 above)
Comment (KGuillemot) [reply vulture_os/portal/views/logon.py:211]: For the moment, we will keep this logger for administrators debug comfort.
It is complicated to change the log level, and info is the default.

File: vulture_os/portal/views/logon.py (hunk 4 above)
Comment (KGuillemot) [reply vulture_os/portal/views/logon.py]: Logger removed, thank you.

File: vulture_os/portal/views/logon.py (hunk 9 above)
Comment (KGuillemot) [reply vulture_os/portal/views/logon.py:642]: I don't think so.

File: vulture_os/services/frontend/models.py (hunk 10 above)
Comment (KGuillemot) [reply vulture_os/services/frontend/models.py:1487]: Yes it is right now, you're right. But the function is not useless, just unused.
We will use this function in the futur.

File: vulture_os/services/frontend/models.py (hunk 11 above)
Comment (KGuillemot) [reply vulture_os/services/frontend/models.py:1605]: That can't append.

File: vulture_os/toolkit/auth/ldap_client.py (hunk 12 above)
Comment (KGuillemot) [reply vulture_os/toolkit/auth/ldap_client.py]: Logger set as debug level.

**Review by frikilax — APPROVED** [review]

**Pull request merged**